src/piecewise.jl

```diff
@@ -0,0 +1,215 @@
+struct Domain{L, R, T, S}
+    lo::T
+    hi::S
+
+    function Domain{L, R, T, S}(lo::T, hi::S) where {L, R, T, S}
+        (!(L in (:open, :closed)) || !(R in (:open, :closed))) && throw(ErrorException(
+            "Domain bound must be either :open or :closed, got $L and $R instead"
+        ))
+        return new{L, R, T, S}(lo, hi)
+    end
+end
+
+Domain{L, R}(lo::T, hi::S) where {T, S, L, R} = Domain{L, R, T, S}(lo, hi)
+Domain(lo, hi) = Domain{:open, :closed}(lo, hi)
+Domain(lo::Tuple, hi::Tuple) = Domain{lo[2], hi[2]}(lo[1], hi[1])
+Domain(X::Interval) = Domain{:closed, :closed}(inf(X), sup(X))
+Domain() = Domain{:open, :open}(Inf, -Inf)
+
+lowerbound(x::Domain{L, R}) where {L, R} = (x.lo, L)
+upperbound(x::Domain{L, R}) where {L, R} = (x.hi, R)
+
+inf(x::Domain) = x.lo
+sup(x::Domain) = x.hi
+
+"""
+    rightof(val, lowerbound)
+
+Determine if a value is on the right of a lower bound.
+"""
+function rightof(x::Real, (val, bound))
+    bound == :closed && return val <= x
+    return val < x
+end
+
+function rightof((val1, bound1), (val2, bound2))
+    if val1 < val2 || (val1 == val2 && bound1 == bound2 == :closed)
+        return false
+    end
+
+    return true
+end
+
+"""
+    leftof(val, upperbound)
+
+Determine if a value is on the left of an upper bound.
+"""
+function leftof(x::Real, (val, bound))
+    bound == :closed && return x <= val
+    return x < val
+end
+
+function leftof((val1, bound1), (val2, bound2))
+    if val2 < val1 || (val1 == val2 && bound1 == bound2 == :closed)
+        return false
+    end
+
+    return true
+end
+
+function leftof(d1::Domain, d2::Domain)
+    val1, bound1 = upperbound(d1)
+    val2, bound2 = lowerbound(d2)
+
+    val1 == val2 && return !(bound1 == bound2 == :closed)
+    return val1 < val2
+end
+
+Base.in(x::Real, domain::Domain) = rightof(x, lowerbound(domain)) && leftof(x, upperbound(domain))
+
+function Base.intersect(d1::Domain, d2::Domain)
+    left = max(lowerbound(d1), lowerbound(d2))
+    right = min(upperbound(d1), upperbound(d2))
+
+    left > right && return Domain()
+    return Domain(left, right)
+end
+
+function Base.isempty(domain::Domain)
+    lo, lobound = lowerbound(domain)
+    hi, hibound = upperbound(domain)
+
+    lo == hi && return !(lobound == hibound == :closed)
+    return lo > hi
+end
+
+
+struct Constant{T}
+    value::T
+end
+
+(constant::Constant)(::Any) = constant.value
+(constant::Constant)(::Interval) = interval(constant.value)
+
+# TODO Proper type parameters
+struct Piecewise
+    domain::Vector
+    fs
+    continuity::Vector{Int}
+    singularities::Vector
+end
```

Comment:
You probably already thought about it, but I would have imagined using `Tuple` here instead of `Vector`. I think these are good below length 16, or 32?

But maybe an implementation like
```Julia
struct Piecewise{N,T<:Tuple,S<:Tuple,R<:Tuple}
     domains :: T
     functions :: S
     continuity_orders :: NTuple{N,Int}
     singularities :: R
 end
```
is messy, or can hurt the performance?

Reply:
We can use tuples. It's just a bit messy because domains and functions can different types, so the type has whole tuple type as type parameters. But that's mostly a technicality.

Reply:
Indeed, I let you decide what you think is best 🙂